### PR TITLE
Updated Opticflow Module to new vision pipeline

### PR DESCRIPTION
**Pull request by knmcguire**
Updated Opticflow Module to new vision pipeline, 

- Made some adjustments and fixes to the code itself
- Tested it on both Bebop and ARdrone2's cameras (first it only worked on the ARdrone)
- Updated airframe files

### CODE REVIEW

**Comment by flixr** [comment]
Can you please coordinate this with @rmeertens who is currently working on #1648

**Comment by rmeertens** [comment]
Doesn't really matter, the first one to be merged doesn't have to change the register function ;)

**Comment by knmcguire** [comment]
Yes that is what we (me and rmeertens) have agreed upon eachother. So if this get merged first, he will change it and visaversa.

**Comment by flixr** [comment]
Ok.
I see you removed the mutex, but I think it is still needed to protect the data (`opticflow_state` and `opticflow_result`) between the function called in the thread and the usage outside or you can end up with out-of-sync data...

**Comment by knmcguire** [comment]
I'm afraid that even with how it was, calling the mutex to collect the state in between, still caused out of sync data with the images. In the old code, with the ARdrone, the state info is about 0.1 ms out of sync with the images already. So I wonder if it would make any difference. 

What I can do is, return the state_temp and result_temp (inside the thread) and copy that data in opticflow_module_run in opticflow_state and opticflow_result (based on a bool that states if the data is new or not) , since that function does not run in the vision thread anymore?

**Comment by knmcguire** [comment]
I returned the mutex, it doesn't have much to do with delay, and I see your point. Hope it is okay now

**Comment by wilcoschoneveld** [comment]
@knmcguire Somehow I cannot create a pull request to your fork with the changes we made, but you can find the commits [here](https://github.com/wilcoschoneveld/paparazzi/tree/pull_request_opticflow_bebop).

**Comment by fvantienen** [comment]
@wilcoschoneveld Do you want the commits also in here? Or should I merge this PR first?

File: sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c [hunk 1]
```diff
@@ -187,6 +191,11 @@ void opticflow_calc_init(struct opticflow_t *opticflow, uint16_t w, uint16_t h)
 void calc_fast9_lukas_kanade(struct opticflow_t *opticflow, struct opticflow_state_t *state, struct image_t *img,
                              struct opticflow_result_t *result)
 {
+  if (opticflow->just_switched_method) {
+    printf("OPTICALFLOW: switched to fast9 Lukas Kanade\n");
```
Comment (fvantienen) [inline sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c]: Can you remove this debugging information?

File: sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c [hunk 2]
```diff
@@ -365,11 +373,27 @@ void calc_edgeflow_tot(struct opticflow_t *opticflow, struct opticflow_state_t *
...
 
   // Define Normal variables
   struct edgeflow_displacement_t displacement;
+  displacement.x = malloc(sizeof(int32_t) * img->w);
+  displacement.y = malloc(sizeof(int32_t) * img->h);
+
+  // If the methods just switched to this one, reintialize the
+  // array of edge_hist structure.
+  if (opticflow->just_switched_method == 1) {
+    printf("OPTICALFLOW: switched to EdgeFlow\n");
```
Comment (fvantienen) [inline sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c]: Same here

**Comment by wilcoschoneveld** [comment]
@fvantienen No, @knmcguire already added the changes in this pull request.

**Comment by fvantienen** [comment]
This looks find 👍 ! Well done!

**Pull request merged**